diff --git a/net/ipv4/icmp.c b/net/ipv4/icmp.c
index 2745493d6b4f..23204b030244 100644
--- a/net/ipv4/icmp.c
+++ b/net/ipv4/icmp.c
@@ -561,6 +561,21 @@ static struct rtable *icmp_route_lookup(struct net *net,
 		dst_release(&rt2->dst);
 		rt2 = skb_rtable(skb_in);
 		skb_in->_skb_refdst = orefdst; /* restore old refdst */
+
+		/*
+		 * At this point, fl4_dec.daddr should NOT be local (we
+		 * checked fl4_dec.saddr above). However, a race condition
+		 * may occur if the address is added to the interface
+		 * concurrently. In that case, ip_route_input() returns a
+		 * LOCAL route with dst.output=ip_rt_bug, which must not
+		 * be used for output.
+		 */
+		if (!err && rt2 && rt2->rt_type == RTN_LOCAL) {
+			net_warn_ratelimited("detected local route for %pI4 during ICMP sending, src %pI4\n",
+					     &fl4_dec.daddr, &fl4_dec.saddr);
+			dst_release(&rt2->dst);
+			err = -EINVAL;
+		}
 	}
 
 	if (err)
