diff --git a/net/ipv4/icmp.c b/net/ipv4/icmp.c
index 23e921d313b3..b29e40953e8f 100644
--- a/net/ipv4/icmp.c
+++ b/net/ipv4/icmp.c
@@ -73,6 +73,7 @@
 #include <linux/string.h>
 #include <linux/netfilter_ipv4.h>
 #include <linux/slab.h>
+#include <linux/delay.h>
 #include <net/flow.h>
 #include <net/snmp.h>
 #include <net/ip.h>
@@ -559,6 +560,7 @@ static struct rtable *icmp_route_lookup(struct net *net, struct flowi4 *fl4,
 		}
 		/* Ugh! */
 		orefdst = skb_dstref_steal(skb_in);
+		mdelay(1000);
 		err = ip_route_input(skb_in, fl4_dec.daddr, fl4_dec.saddr,
 				     dscp, rt2->dst.dev) ? -EINVAL : 0;
 
