diff --git a/net/ipv4/icmp.c b/net/ipv4/icmp.c
index 2745493d6b4f..39de61e3e956 100644
--- a/net/ipv4/icmp.c
+++ b/net/ipv4/icmp.c
@@ -88,6 +88,7 @@
 #include <linux/timer.h>
 #include <linux/init.h>
 #include <linux/uaccess.h>
+#include <linux/delay.h>
 #include <net/checksum.h>
 #include <net/xfrm.h>
 #include <net/inet_common.h>
@@ -555,6 +556,7 @@ static struct rtable *icmp_route_lookup(struct net *net,
 		/* Ugh! */
 		orefdst = skb_in->_skb_refdst; /* save old refdst */
 		skb_dst_set(skb_in, NULL);
+		mdelay(1000);
 		err = ip_route_input(skb_in, fl4_dec.daddr, fl4_dec.saddr,
 				     RT_TOS(tos), rt2->dst.dev);
 
